src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,327 @@
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
                 }
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
+
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source compaction frontier. This gives permission to the
+    /// reclock operator to compact and consolidate its internal state accordingly.
+    ///
+    /// If the given source since frontier results in a change of the dest since frontier it will
+    /// be returned to the caller who can then use it to downgrade its capability.
+    pub async fn compact(
+        &mut self,
+        source_since: &HashMap<PartitionId, MzOffset>,
+    ) -> Option<AntichainRef<'_, Timestamp>> {
+        // Translate the SourceTime compaction frontier into a DestTime compaction frontier
+        let new_since = source_since
+            .iter()
+            .map(|(pid, part_since)| {
+                self.partition_uppers(pid)
+                    .find(|(_ts, offset)| part_since < offset)
+                    .expect("attempt to compact past upper")
+                    .0
+            })
+            .min()
+            .map(Antichain::from_elem)
+            .unwrap_or_else(|| self.since.clone());
+
+        if PartialOrder::less_than(&self.since, &new_since) {
+            for bindings in self.remap_trace.values_mut() {
+                // Compact the remap trace according to the computed frontier
+                for (timestamp, _) in bindings.iter_mut() {
+                    timestamp.advance_by(new_since.borrow());
+                }
+                // And then consolidate
+                consolidation::consolidate(bindings);
+            }
+            self.since = new_since;
+            self.read_handle.downgrade_since(self.since.clone()).await;
+            Some(self.since.borrow())
+        } else {
+            None
+        }
+    }
+
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
+        loop {
+            let mut new_ts = (self.now)();
+            new_ts -= new_ts % self.update_interval_ms;
+            if self.upper.less_than(&new_ts) {
+                break new_ts;
+            }
+            let sleep_ms = self.update_interval_ms - (new_ts % self.update_interval_ms);
+            tokio::time::sleep(Duration::from_millis(sleep_ms)).await;
+        }
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Returns an iterator of timestamp bindings for a given partition
+    fn partition_uppers(&self, pid: &PartitionId) -> PartitionUppers {
```

Comment:
Hmm I don't think I understand this naming, which is making me worried I'm missing something.  Why the `_uppers` suffix?

Reply:
You can think of the remap collection as storing an upper frontier (albeit not stored as an `Antichain`) as it advances through time. So when the collection contains the following for some partition P:

```
Timestamp(1000), Diff(10) (i.e MzOffset(10))
Timestamp(2000), Diff(10) (i.e MzOffset(20))
```

It means that at time `1000` the upper frontier of partition `P` was 10, which means that all messages at **offsets < 10** should be timestamped with `1000` and at **offsets < 20** at time `2000`. Crucially we don't know yet what timestamp offset 20 will be assigned at.

The reason this is an upper (which turns into a `<` relation instead of a `<=`) is because not having an entry in the `HashMap` doesn't mean that an upper frontier doesn't exist. We might haven't seen messages from partition `P2` yet, and so it's not in the map yet, but nevertheless its `MzOffset` is as if it was set to zero. If we didn't have this situation where the list of partitions cannot be known in advance we could be truly using an `Antichain` to store it.